subprojects/core/src/test/groovy/org/gradle/api/internal/AbstractDomainObjectCollectionSpec.groovy

```diff
@@ -598,4 +598,19 @@ abstract class AbstractDomainObjectCollectionSpec<T> extends Specification {
         0 * action.execute(a)
         1 * action.execute(b)
     }
+
+    def "can mutate the container inside a configureEach action"() {
+        given:
+        container.add(a)
+        container.add(b)
+        container.add(c)
+
+        when:
+        container.configureEach {
+            container.add(d)
```

Comment:
I think this is something we want to prevent eventually (modifying the container by creating/registering an object from a lazy configuration action).

Reply:
At that point we're going to need a proper way of registering new elements in response to the registration of another element. Something like `whenObjectRegistered(action: Action<NamedDomainObjectProvider<T>>)` which would allow for controlled mutations and callback serialisation during `action` execution.